react/features/facial-recognition/facialExpressionsWorker.js

```diff
@@ -1,120 +1,164 @@
+ * Contains the last valid face bounding box (passes threshold validation) which was sent to the main process.
  */
-const window = {
-    screen: {
-        width: 1280,
-        height: 720
+let lastValidFaceBox;
+
+const detectFaceBox = async ({ imageTensor, threshold }) => {
+    const detections = await faceapi.detectAllFaces(imageTensor, new faceapi.TinyFaceDetectorOptions());
+
+    if (!detections.length) {
+        return;
+    }
+
+    const faceBox = {
+        // normalize to percentage based
+        left: Math.round(Math.min(...detections.map(d => d.relativeBox.left)) * 100),
+        right: Math.round(Math.max(...detections.map(d => d.relativeBox.right)) * 100),
+        top: Math.round(Math.min(...detections.map(d => d.relativeBox.top)) * 100),
+        bottom: Math.round(Math.max(...detections.map(d => d.relativeBox.bottom)) * 100)
+    };
+
+    if (lastValidFaceBox && Math.abs(lastValidFaceBox.left - faceBox.left) < threshold) {
+        return;
     }
+
+    lastValidFaceBox = faceBox;
+
+    self.postMessage({
+        type: FACE_BOX_MESSAGE,
+        value: faceBox
+    });
 };
 
-onmessage = async function(message) {
-    switch (message.data.type) {
-    case INIT_WORKER : {
-        modelsURL = message.data.url;
-        if (message.data.windowScreenSize) {
-            window.screen = message.data.windowScreenSize;
-        }
-        break;
+const detectFaceExpressions = async ({ imageTensor }) => {
+    const detections = await faceapi.detectSingleFace(
+        imageTensor,
+        new faceapi.TinyFaceDetectorOptions()
+    ).withFaceExpressions();
+
+    const facialExpression = detections?.expressions.asSortedArray()[0].expression;
+
+    if (facialExpression) {
+        self.postMessage({
+            type: FACIAL_EXPRESSION_MESSAGE,
+            value: facialExpression
+        });
     }
+};
 
-    case SET_TIMEOUT : {
-        if (!message.data.imageBitmap || !modelsURL) {
-            self.postMessage({
-                type: FACIAL_EXPRESSION_MESSAGE,
-                value: null
-            });
-        }
+const detect = async ({ image, threshold }) => {
+    detectionInProgress = true;
+    faceapi.tf.engine().startScope();
 
-        // the models are loaded
-        if (!modelsLoaded) {
-            await faceapi.loadTinyFaceDetectorModel(modelsURL);
-            await faceapi.loadFaceExpressionModel(modelsURL);
-            modelsLoaded = true;
+    const imageTensor = faceapi.tf.browser.fromPixels(image);
+
+    if (faceDetectionTypes.includes(DETECTION_TYPES.FACE_BOX)) {
+        await detectFaceBox({
+            imageTensor,
+            threshold
+        });
+    }
+
+    if (faceDetectionTypes.includes(DETECTION_TYPES.FACE_EXPRESSIONS)) {
+        await detectFaceExpressions({ imageTensor });
+    }
+
+    faceapi.tf.engine().endScope();
+    detectionInProgress = false;
+};
+
+const init = async ({ baseUrl, detectionTypes }) => {
+    initInProgress = true;
```

Comment:
Ditto.

Reply:
init is only called once, if it's called twice, we have a bigger problem

Reply:
Then maybe we can drop this to make it a bit less confusing for readers? As it stands it looks like it can be called multiple times, and we guard against it.

Reply:
actually I can make use of other flags for that, ill remove this